src/v/cloud_storage/cache_service.cc

```diff
@@ -205,24 +258,46 @@ ss::future<> cache::put(
           tmp_extension));
     }
     auto dir_path = (_cache_dir / key).remove_filename();
-    co_await ss::recursive_touch_directory(dir_path.string());
 
-    // tmp file is used to protect against concurrent writes to the same file.
-    // One tmp file is written only once by one thread. tmp file should not be
-    // read directly. _cnt is an atomic counter that ensures the uniqueness of
-    // names for tmp files within one shard, while shard_id ensures uniqueness
-    // across multiple shards.
+    // tmp file is used to protect against concurrent writes to the same
+    // file. One tmp file is written only once by one thread. tmp file
+    // should not be read directly. _cnt is an atomic counter that
+    // ensures the uniqueness of names for tmp files within one shard,
+    // while shard_id ensures uniqueness across multiple shards.
     auto tmp_filename = std::filesystem::path(ss::format(
       "{}_{}_{}{}",
       filename.native(),
       ss::this_shard_id(),
       (++_cnt),
       tmp_extension));
-    auto flags = ss::open_flags::wo | ss::open_flags::create
-                 | ss::open_flags::exclusive;
 
-    auto tmp_cache_file = co_await ss::open_file_dma(
-      (dir_path / tmp_filename).native(), flags);
+    ss::file tmp_cache_file;
+    bool try_open_file = true;
+
+    while (try_open_file) {
```

Comment:
can this be achieved with a simple break statement?

Reply:
I think I would need to add a `continue` in a catch block and `break` after the catch block. I think that would be less readable.

Reply:
I was thinking of something like this:
```suggestion
while (true) {
    co_await ss::recursive_touch_directory(dir_path.string());
    // recursive_delete_empty_directory may delete dir_path before we open
    // file, in this case we recreate dir_path and try again
    try {
        auto flags = ss::open_flags::wo | ss::open_flags::create
                     | ss::open_flags::exclusive;

        tmp_cache_file = co_await ss::open_file_dma(
          (dir_path / tmp_filename).native(), flags);
        break;
    } catch (std::filesystem::filesystem_error& e) {
        if (e.code() == std::errc::no_such_file_or_directory) {
            vlog(
              cst_log.debug,
              "Couldn't open {}, gonna retry",
              (dir_path / tmp_filename).native());
        } else {
            throw;
        }
    }
}
```
I think it is it the same?